File: plant_catalog.cpp
// plant_catalog.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <map>
#include <ctime>
#include <iomanip>
#include <sstream>
#include <algorithm>
#include <json/json.h> // using jsoncpp

using namespace std;

const string DATA_FILE = "plants.json";

struct Plant {
    string name;
    string type;
    int frequency;
    string last_watered;
};

class Catalog {
private:
    vector<Plant> plants;

    string today() {
        time_t t = time(nullptr);
        tm* now = localtime(&t);
        char buf[11];
        strftime(buf, sizeof(buf), "%Y-%m-%d", now);
        return string(buf);
    }

    int days_since(const string& date) {
        tm tm1 = {};
        tm tm2 = {};
        strptime(date.c_str(), "%Y-%m-%d", &tm1);
        strptime(today().c_str(), "%Y-%m-%d", &tm2);
        time_t t1 = mktime(&tm1);
        time_t t2 = mktime(&tm2);
        return (t2 - t1) / 86400;
    }

    string next_water(const string& date, int freq) {
        tm tm1 = {};
        strptime(date.c_str(), "%Y-%m-%d", &tm1);
        tm1.tm_mday += freq;
        mktime(&tm1);
        char buf[11];
        strftime(buf, sizeof(buf), "%Y-%m-%d", &tm1);
        return string(buf);
    }

public:
    Catalog() { load(); }

    void load() {
        ifstream ifs(DATA_FILE);
        if (!ifs) return;
        Json::Value root;
        ifs >> root;
        for (const auto& item : root) {
            Plant p;
            p.name = item["name"].asString();
            p.type = item["type"].asString();
            p.frequency = item["frequency"].asInt();
            p.last_watered = item["last_watered"].asString();
            plants.push_back(p);
        }
    }

    void save() {
        Json::Value root(Json::arrayValue);
        for (const auto& p : plants) {
            Json::Value item;
            item["name"] = p.name;
            item["type"] = p.type;
            item["frequency"] = p.frequency;
            item["last_watered"] = p.last_watered;
            root.append(item);
        }
        ofstream ofs(DATA_FILE);
        ofs << root.toStyledString();
    }

    void addPlant(const string& name, const string& type, int frequency, const string& last) {
        Plant p;
        p.name = name;
        p.type = type;
        p.frequency = frequency;
        p.last_watered = last.empty() ? today() : last;
        plants.push_back(p);
        save();
        cout << "\033[32m🌱 Растение '" << name << "' добавлено.\033[0m" << endl;
    }

    void waterPlant(const string& name) {
        for (auto& p : plants) {
            if (p.name == name) {
                p.last_watered = today();
                save();
                cout << "\033[32m💧 Растение '" << name << "' полито сегодня.\033[0m" << endl;
                return;
            }
        }
        cout << "\033[31m❌ Растение '" << name << "' не найдено.\033[0m" << endl;
    }

    void removePlant(const string& name) {
        for (auto it = plants.begin(); it != plants.end(); ++it) {
            if (it->name == name) {
                plants.erase(it);
                save();
                cout << "\033[33m🗑️ Растение '" << name << "' удалено.\033[0m" << endl;
                return;
            }
        }
        cout << "\033[31m❌ Растение '" << name << "' не найдено.\033[0m" << endl;
    }

    void listPlants(const string& filter) {
        vector<Plant> list = plants;
        if (!filter.empty()) {
            list.clear();
            for (const auto& p : plants) {
                if (p.type == filter) list.push_back(p);
            }
            if (list.empty()) {
                cout << "\033[33m❌ Растения типа '" << filter << "' не найдены.\033[0m" << endl;
                return;
            }
        }
        if (list.empty()) {
            cout << "\033[33m📭 Каталог пуст.\033[0m" << endl;
            return;
        }
        cout << "\033[36m🌿 Каталог растений:\033[0m" << endl;
        for (const auto& p : list) {
            int days = days_since(p.last_watered);
            string status, color;
            if (days <= p.frequency) {
                status = "✅"; color = "\033[32m";
            } else if (days <= p.frequency * 1.5) {
                status = "⚠️"; color = "\033[33m";
            } else {
                status = "🚨"; color = "\033[31m";
            }
            string next = next_water(p.last_watered, p.frequency);
            cout << color << status << " " << p.name << " (" << p.type << ") - полив каждые " << p.frequency << " дн., след. полив: " << next << "\033[0m" << endl;
        }
    }

    void exportJSON(const string& filename) {
        Json::Value root(Json::arrayValue);
        for (const auto& p : plants) {
            Json::Value item;
            item["name"] = p.name;
            item["type"] = p.type;
            item["frequency"] = p.frequency;
            item["last_watered"] = p.last_watered;
            root.append(item);
        }
        ofstream ofs(filename);
        ofs << root.toStyledString();
        cout << "\033[32m📄 Экспортировано в " << filename << " (JSON)\033[0m" << endl;
    }

    void exportCSV(const string& filename) {
        ofstream ofs(filename);
        ofs << "name,type,frequency,last_watered\n";
        for (const auto& p : plants) {
            ofs << p.name << "," << p.type << "," << p.frequency << "," << p.last_watered << "\n";
        }
        cout << "\033[32m📄 Экспортировано в " << filename << " (CSV)\033[0m" << endl;
    }

    void exportTXT(const string& filename) {
        ofstream ofs(filename);
        for (const auto& p : plants) {
            ofs << p.name << " | " << p.type << " | " << p.frequency << " дн. | последний полив: " << p.last_watered << "\n";
        }
        cout << "\033[32m📄 Экспортировано в " << filename << " (TXT)\033[0m" << endl;
    }
};

int main(int argc, char* argv[]) {
    string add, type, last, filter, water, remove, json, csv, txt;
    int frequency = 0;
    bool list = false;

    for (int i = 1; i < argc; ++i) {
        string arg = argv[i];
        if (arg == "--add" && i+1 < argc) add = argv[++i];
        else if (arg == "--type" && i+1 < argc) type = argv[++i];
        else if (arg == "--frequency" && i+1 < argc) frequency = stoi(argv[++i]);
        else if (arg == "--last" && i+1 < argc) last = argv[++i];
        else if (arg == "--list") list = true;
        else if (arg == "--filter" && i+1 < argc) filter = argv[++i];
        else if (arg == "--water" && i+1 < argc) water = argv[++i];
        else if (arg == "--remove" && i+1 < argc) remove = argv[++i];
        else if (arg == "--export-json" && i+1 < argc) json = argv[++i];
        else if (arg == "--export-csv" && i+1 < argc) csv = argv[++i];
        else if (arg == "--export-txt" && i+1 < argc) txt = argv[++i];
    }

    Catalog catalog;
    if (!add.empty()) {
        if (type.empty() || frequency == 0) {
            cerr << "❌ Для добавления растения требуются --type и --frequency" << endl;
            return 1;
        }
        catalog.addPlant(add, type, frequency, last);
    } else if (list) {
        catalog.listPlants(filter);
    } else if (!water.empty()) {
        catalog.waterPlant(water);
    } else if (!remove.empty()) {
        catalog.removePlant(remove);
    } else if (!json.empty()) {
        catalog.exportJSON(json);
    } else if (!csv.empty()) {
        catalog.exportCSV(csv);
    } else if (!txt.empty()) {
        catalog.exportTXT(txt);
    } else {
        cout << "Используйте --help для справки." << endl;
    }
    return 0;
}

File: plant_catalog.py

#!/usr/bin/env python3
# plant_catalog.py
import argparse
import json
import csv
import sys
from datetime import datetime, timedelta
from colorama import init, Fore, Style

init(autoreset=True)

DATA_FILE = "plants.json"

class PlantCatalog:
    def __init__(self):
        self.plants = []
        self.load()

    def load(self):
        try:
            with open(DATA_FILE, 'r', encoding='utf-8') as f:
                self.plants = json.load(f)
        except (FileNotFoundError, json.JSONDecodeError):
            self.plants = []

    def save(self):
        with open(DATA_FILE, 'w', encoding='utf-8') as f:
            json.dump(self.plants, f, ensure_ascii=False, indent=2)

    def add_plant(self, name, plant_type, frequency, last_watered=None):
        if last_watered is None:
            last_watered = datetime.now().date().isoformat()
        plant = {
            "name": name,
            "type": plant_type,
            "frequency": frequency,
            "last_watered": last_watered
        }
        self.plants.append(plant)
        self.save()
        print(f"🌱 Растение '{name}' добавлено.")

    def water_plant(self, name):
        for plant in self.plants:
            if plant["name"].lower() == name.lower():
                plant["last_watered"] = datetime.now().date().isoformat()
                self.save()
                print(f"💧 Растение '{name}' полито сегодня.")
                return
        print(f"❌ Растение '{name}' не найдено.")

    def remove_plant(self, name):
        for i, plant in enumerate(self.plants):
            if plant["name"].lower() == name.lower():
                del self.plants[i]
                self.save()
                print(f"🗑️ Растение '{name}' удалено.")
                return
        print(f"❌ Растение '{name}' не найдено.")

    def get_status(self, plant):
        last = datetime.fromisoformat(plant["last_watered"]).date()
        today = datetime.now().date()
        days_since = (today - last).days
        freq = plant["frequency"]
        if days_since <= freq:
            return "✅", Fore.GREEN
        elif days_since <= freq * 1.5:
            return "⚠️", Fore.YELLOW
        else:
            return "🚨", Fore.RED

    def list_plants(self, filter_type=None):
        if not self.plants:
            print("📭 Каталог пуст.")
            return
        plants = self.plants
        if filter_type:
            plants = [p for p in self.plants if p["type"].lower() == filter_type.lower()]
            if not plants:
                print(f"❌ Растения типа '{filter_type}' не найдены.")
                return
        print("🌿 Каталог растений:")
        for plant in plants:
            status, color = self.get_status(plant)
            next_water = datetime.fromisoformat(plant["last_watered"]).date() + timedelta(days=plant["frequency"])
            print(f"{color}{status} {plant['name']} ({plant['type']}) - полив каждые {plant['frequency']} дн., след. полив: {next_water.isoformat()}{Style.RESET_ALL}")

    def export_json(self, filename):
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(self.plants, f, ensure_ascii=False, indent=2)
        print(f"📄 Экспортировано в {filename} (JSON)")

    def export_csv(self, filename):
        with open(filename, 'w', newline='', encoding='utf-8') as f:
            writer = csv.DictWriter(f, fieldnames=["name", "type", "frequency", "last_watered"])
            writer.writeheader()
            writer.writerows(self.plants)
        print(f"📄 Экспортировано в {filename} (CSV)")

    def export_txt(self, filename):
        with open(filename, 'w', encoding='utf-8') as f:
            for plant in self.plants:
                f.write(f"{plant['name']} | {plant['type']} | {plant['frequency']} дн. | последний полив: {plant['last_watered']}\n")
        print(f"📄 Экспортировано в {filename} (TXT)")

def main():
    parser = argparse.ArgumentParser(description="Каталог растений (полив)")
    parser.add_argument("--add", help="Добавить растение")
    parser.add_argument("--type", help="Тип растения")
    parser.add_argument("--frequency", type=int, help="Частота полива (дни)")
    parser.add_argument("--last", help="Дата последнего полива (YYYY-MM-DD)")
    parser.add_argument("--list", action="store_true", help="Показать все растения")
    parser.add_argument("--filter", help="Фильтр по типу")
    parser.add_argument("--water", help="Отметить растение политым")
    parser.add_argument("--remove", help="Удалить растение")
    parser.add_argument("--export-json", help="Экспорт в JSON")
    parser.add_argument("--export-csv", help="Экспорт в CSV")
    parser.add_argument("--export-txt", help="Экспорт в TXT")
    args = parser.parse_args()

    catalog = PlantCatalog()

    if args.add:
        if not args.type or not args.frequency:
            print("❌ Для добавления растения требуются --type и --frequency")
            sys.exit(1)
        catalog.add_plant(args.add, args.type, args.frequency, args.last)
    elif args.list:
        catalog.list_plants(args.filter)
    elif args.water:
        catalog.water_plant(args.water)
    elif args.remove:
        catalog.remove_plant(args.remove)
    elif args.export_json:
        catalog.export_json(args.export_json)
    elif args.export_csv:
        catalog.export_csv(args.export_csv)
    elif args.export_txt:
        catalog.export_txt(args.export_txt)
    else:
        parser.print_help()

if __name__ == "__main__":
    main()
